xapian-core/common/io_utils.cc

```diff
@@ -174,7 +174,7 @@ io_pread(int fd, char * p, size_t n, off_t o, size_t min)
 	    // We get EINTR if the syscall was interrupted by a signal.
 	    // In this case we should retry the read.
 	    if (errno == EINTR) continue;
-	    throw Xapian::DatabaseError("Error reading database", errno);
+	    throw Xapian::DatabaseClosedError("Error reading database", errno);
```

Comment:
This is reporting an error trying to read a block from the database, not reporting an attempt to do something to a closed database.

If the database is closed then the file descriptor is set to `-2` to indicate this and that should be checked before we get here.

Reply:
thanks fixed.

Reply:
Sorry, I wasn't clear about this - by "before we get here" I really meant "at a higher level" rather than at the start of this function.

The changes I made on master in 2a980274452b642739f799f0c066d9cdfb3048da to fix the dangling closed fd issue should mean that `fd == -2` never happens here, and indeed I can revert all the changes to `io_utils.cc` made on your branch and the tests all still pass.

And should that assumption be violated, the syscalls here will fail with `errno` set to `EBADFD` we should see a testcase failure rather than it getting quietly handled.

Don't worry about changing that - I can easily merge with those changes reverted.